base/bundle.hpp

```diff
@@ -0,0 +1,140 @@
+#include <vector>
+
+#include "base/macros.hpp"
+#include "base/not_null.hpp"
+#include "base/status.hpp"
+
+namespace principia {
+namespace base {
+
+// Functions that can be cooperatively aborted should check |AbortRequested()|
+// when appropriate.
+extern thread_local std::function<bool()> AbortRequested;
+
+class Bundle {
+  class Task;
+ public:
+  class TaskHandle {
+   public:
+    TaskHandle() = default;
+    TaskHandle(TaskHandle const&) = delete;
+    TaskHandle(TaskHandle&&) = default;
+    TaskHandle& operator=(TaskHandle const&) = delete;
+    TaskHandle& operator=(TaskHandle&&) = default;
+
+   private:
+    explicit TaskHandle(std::shared_ptr<Task> task);
+    std::shared_ptr<Task> task_;
+    friend class Bundle;
+  };
+
+  explicit Bundle(int workers);
+
+  // Returns the first non-OK status encountered.  All worker threads are
+  // joined; no calls to |Add| or |Join| may follow this call.
+  Status Join();
+
+  // The |TaskHandle| returned by |Add| may be used in a call to |Abort| or
+  // |JoinTask|.  |this| shares ownership of the |Task| with the handle until
+  // the |Task| is |Done|.
+  TaskHandle Add(std::function<Status()> task_body);
+
+  // Removes the given |task| from the queue if it has not started, or instructs
+  // it to cooperatively abort if it is running.
+  void Abort(TaskHandle const& task);
+
+  // Waits until the given |task| completes, and returns its status.  Takes
+  // ownership of |task|, so that |*task->task_| is destroyed when this function
+  // returns.
+  Status JoinTask(TaskHandle const task);
+
+ private:
+  class Task {
+   public:
+    // All state transitions set additional bits; No state transition clears a
+    // bit.
+    enum ExecutionState : std::uint8_t {
+      // |this| is in the queue waiting for a thread to execute it.
+      Inactive = 0b000,
+      // |body_()| is being executed.
+      Active = 0b001,
+      // |body_()| has returned or |this| has been dequeued from the |Inactive|
+      // state by an |Abort|.
+      Terminated = 0b010,
+      // |Abort| has been called.
+      Aborted = 0b100,
+    };
+
+    Task(not_null<Bundle*> bundle, std::function<Status()> body);
+
+    std::uint8_t execution_state() const;
+
+    // |bundle_->lock_| must be held by the caller.  Requires
+    // |execution_state() == Inactive|.
+    void set_queue_it(std::list<std::shared_ptr<Task>>::iterator it);
+
+    // No prerequisites.  If |Inactive|, |body_()| never runs.  If |Active|,
+    // a cooperative abort is requested.  Sets the |Aborted| bit.
+    void Abort();
+
+    // Requires |execution_state() == Inactive|.  Sets the |Active| bit, and
+    // calls |body_()|.  When |body_()| returns, sets the |Terminated| bit, and
+    // notifies |terminated_|.
+    void Execute();
+
+    // Requires |execution_state() & Terminated|.
+    // Returns the status returned by |body()|, or a status with
+    // |Error::ABORTED| if |Abort| was called while |Inactive|.
+    Status status() const;
```

Comment:
It would be more logical to return the status in `wait_for_termination` as calling it at any other point is incorrect anyway.